[ot] hw/opentitan: ot_spi_host: rework SPI Host

rivos-eblot opened the pull request:
* fix STATUS.active bit
* fix command FIFO management
* add basic SPI bus clock pacing
* improve FSM scheduling

All SPI Host tests from `sw/device/tests` now pass
```
+---------------------------------+--------+----------+--------+-------+
| Name                            | Result |     Time | Icount | Error |
+=================================+========+==========+========+=======+
| spi_host_irq_test               |   PASS |  6668 ms |      6 |       |
+---------------------------------+--------+----------+--------+-------+
| spi_host_macronix1Gb_flash_test |   PASS | 13669 ms |      6 |       |
+---------------------------------+--------+----------+--------+-------+
| spi_host_smoketest              |   PASS |  6786 ms |      6 |       |
+---------------------------------+--------+----------+--------+-------+
| spi_host_winbond_flash_test     |   PASS | 11870 ms |      6 |       |
+---------------------------------+--------+----------+--------+-------+
```

See https://github.com/lowRISC/qemu/issues/129 for details.

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: add PCLK property

Ideally, this should be implemented as a irq signal to receive
dynamic clock settings from the clock manager. Until ot_clkmgr supports
this feature, use a simple property.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/riscv: ot_earlgrey: configure SPI host clocks

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: fix STATUS.active bit behavior

Documentation was misleading: "the command is complete when STATUS.ACTIVE goes low."
It turns out STATUS.active goes low when the last pushed command is completed,
in other words, STATUS.active is not released as soon as a command is complete.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: add experimental clock pacing

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: add missing function name to log messages

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: fix internal reset

Update SPI Host doc clarifies that registers are -not- reset upon a SW reset.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/riscv: ot_darjeeling: configure SPI host clock

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: remove bottom half that has been deprecated

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: remove duplicate code

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: resume command execution even on error

Observing Verilator waves, it seems nothing precludes the SPI HOST FSM
from executing on error.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: rework command FIFO management

Observing Verilator waves, the command FIFO only contains next command,
not the one that is being executed. In other words, there can be 5 in-fly commands,
not 4, as 4 are stored in the command FIFO + 1 is being executed.

Remove CS channel limitation (from 8 to 32 bits) and command tracking value
(from 16 to 32 bits)

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: restore reduced status trace configuration

Restore configuration that got unexpectedly discarded in "rework SPI event management" commit

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: deprecated fsm.active bit

Active command status provides the same information.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: add command status to ot_spi_host_trace_status

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: refactor FSM scheduling

Command completion is handled in a dedicated function.
Add a short delay on command push to yield execution back to the vCPU
before kicking the FSM.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: add missing const qualifiers

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: fix TX/RX stall management on new command

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: ensure invalid CSID is a no-op on SPI I/F

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: fix comment style

Remove deprecated Doxygen-like C comments, use QEMU style.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] docs/opentitan: earlgrey & darjeeling: update SPI Host documentation

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

loiclefort commented on hw/opentitan/ot_spi_host.c:
```diff
@@ -1162,7 +1138,8 @@ static void ot_spi_host_io_write(void *opaque, hwaddr addr, uint64_t val64,
             F_COMMAND_DIRECTION[FIELD_EX32(slot.command, COMMAND, DIRECTION)],
             F_COMMAND_SPEED[FIELD_EX32(slot.command, COMMAND, SPEED)],
-            (uint32_t)csid, (bool)FIELD_EX32(slot.command, COMMAND, CSAAT),
+            (uint32_t)s->regs[R_CSID] & 0xffu,
```
I think the cast is not required anymore

rivos-eblot replied:
Nor the mask, I though I had fixed it, thanks

rivos-eblot replied:
fixed

loiclefort commented on hw/opentitan/ot_spi_host.c:
```diff
@@ -1235,9 +1232,9 @@ static const MemoryRegionOps ot_spi_host_ops = {
-    DEFINE_PROP_UINT32("bus-num", OtSPIHostState, bus_num, 0),
-    DEFINE_PROP_UINT32("pclk", OtSPIHostState, pclk, 0),
+    DEFINE_PROP_UINT32("num-cs", OtSPIHostState, num_cs, 1u),
+    DEFINE_PROP_UINT32("bus-num", OtSPIHostState, bus_num, 0u),
+    DEFINE_PROP_UINT32("pclk", OtSPIHostState, pclk, 0u),
```
I think we usually don't put `u` suffix on zero constants

loiclefort replied:
Let's keep it like this. As discussed offline, our convention originated in a compiler bug that is not relevant anymore.

loiclefort commented on docs/opentitan/darjeeling.md:
```diff
@@ -247,6 +247,20 @@ To force QEMU to execute an application despite this feature, bypassing the DFT
 
+* `-global ot-spi_host.start-delay=<time>` may be used to change the default SPI Host FSM start
+  delay. This delay is used to yield back the control to the vCPU before kicking of the execution
+  of a new SPI Host command, so that the guest code get a chance to check the statuses of the SPI
```
> the guest code get**s** a chance

(same fix to do in EG doc)

rivos-eblot replied:
fixed

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: remove banked CONFIGOPTS management

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: remove unused constant

OT_SPI_HOST_IRQ_NUM is no longer used.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>

rivos-eblot committed:
[ot] hw/opentitan: ot_spi_host: fix ResettableAPI usage

When overriding the parent's reset implementation, the child is responsible
to call its parent's implementation.

Signed-off-by: Emmanuel Blot <eblot@rivosinc.com>
